src/backend/d3d12/BindGroupD3D12.cpp

```diff
@@ -0,0 +1,153 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "BindGroupD3D12.h"
+#include "BindGroupLayoutD3D12.h"
+#include "BufferD3D12.h"
+#include "SamplerD3D12.h"
+#include "TextureD3D12.h"
+
+#include "D3D12Backend.h"
+
+namespace backend {
+namespace d3d12 {
+
+    BindGroup::BindGroup(Device* device, BindGroupBuilder* builder)
+        : BindGroupBase(builder), device(device) {
+
+        const auto* bgl = ToBackend(GetLayout());
+        const auto& layout = bgl->GetBindingInfo();
+
+        uint32_t cbvOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::CBV);
+        uint32_t uavOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::UAV);
+        uint32_t srvOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::SRV);
+        uint32_t samplerOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::Sampler);
+
+        for (size_t binding = 0; binding < layout.mask.size(); ++binding) {
+            if (!layout.mask[binding]) {
+                continue;
+            }
+
+            switch (layout.types[binding]) {
+                case nxt::BindingType::UniformBuffer:
+                    {
+                        auto* view = ToBackend(GetBindingAsBufferView(binding));
+                        descriptors[binding].cbv = view->GetCBVDescriptor();
+                        descriptorOffsets[binding] = cbvOffset++;
+                        cbvUavSrvCount++;
+                    }
+                    break;
+                case nxt::BindingType::StorageBuffer:
+                    {
+                        auto* view = ToBackend(GetBindingAsBufferView(binding));
+                        descriptors[binding].uav = view->GetUAVDescriptor();
+                        descriptorOffsets[binding] = uavOffset++;
+                        cbvUavSrvCount++;
+                    }
+                    break;
+                case nxt::BindingType::SampledTexture:
+                    {
+                        auto* view = ToBackend(GetBindingAsTextureView(binding));
+                        descriptors[binding].srv = view->GetSRVDescriptor();
+                        descriptorOffsets[binding] = srvOffset++;
+                        cbvUavSrvCount++;
+                    }
+                    break;
+                case nxt::BindingType::Sampler:
+                    {
+                        auto* sampler = ToBackend(GetBindingAsSampler(binding));
+                        descriptors[binding].sampler = sampler->GetSamplerDescriptor();
+                        descriptorOffsets[binding] = samplerOffset++;
+                        samplerCount++;
+                    }
+                    break;
+            }
+        }
+    }
+
+    void BindGroup::RecordDescriptors(const DescriptorHeapHandle &cbvUavSrvHeapStart, uint32_t* cbvUavSrvHeapOffset, const DescriptorHeapHandle &samplerHeapStart, uint32_t* samplerHeapOffset) {
```

Comment:
Can we have the serial be passed as an argument and have a GetHeapSerial? This will help with a later refactor that will allow multiple heaps per queue submit.